### PR TITLE
adding no-action test, some i instructions, and single syscall impl

### PR DESCRIPTION
**Pull request by svv232**

### CODE REVIEW

File: o1vm/src/interpreters/riscv32i/mod.rs [hunk 1]
```diff
@@ -1,12 +1,12 @@
...
 // FIXME: the value will be updated when the interpreter is fully
 // implemented. Using a small value for now.
-pub const SCRATCH_SIZE: usize = 20;
+pub const SCRATCH_SIZE: usize = 80;
 
 /// Number of instructions in the ISA
 // FIXME: the value might not be correct. It will be updated when the
 // interpreter is fully implemented.
-pub const INSTRUCTION_SET_SIZE: usize = 47;
+pub const INSTRUCTION_SET_SIZE: usize = 90;
```
Comment (dannywillems) [inline o1vm/src/interpreters/riscv32i/mod.rs:9]: I guess we can set the right value now?

File: o1vm/src/interpreters/riscv32i/mod.rs [hunk 2]
```diff
@@ -1,12 +1,12 @@
 /// The minimal number of columns required for the VM
 // FIXME: the value will be updated when the interpreter is fully
 // implemented. Using a small value for now.
-pub const SCRATCH_SIZE: usize = 20;
+pub const SCRATCH_SIZE: usize = 80;
```
Comment (dannywillems) [inline o1vm/src/interpreters/riscv32i/mod.rs:4]: I guess we can try to set a value the no-action program requires atm?

File: o1vm/src/interpreters/riscv32i/interpreter.rs [hunk 3]
```diff
@@ -1077,8 +1076,157 @@ pub fn interpret_rtype<Env: InterpreterEnv>(_env: &mut Env, _instr: RInstruction
...
+            env.set_next_instruction_pointer(next_instruction_pointer + Env::constant(4u32));
+        }
+        IInstruction::JumpAndLinkRegister => {
+            let addr = env.read_register(&rs1);
+            // jalr:
+            //  t  = pc+4;
+            //  pc = (x[rs1] + sext(offset)) & ∼1; <- NOT NOW
+            //  pc = (x[rs1] + sext(offset)); <- PLEASE FIXME
+            //  x[rd] = t
+            // copying mips for now to match deugger
```
Comment (dannywillems) [inline o1vm/src/interpreters/riscv32i/interpreter.rs:1213]: I guess we can remove this?

File: o1vm/src/interpreters/riscv32i/interpreter.rs [hunk 4]
```diff
@@ -1077,8 +1076,157 @@ pub fn interpret_rtype<Env: InterpreterEnv>(_env: &mut Env, _instr: RInstruction
...
+            + v0
+    };
+
+    /* fetch opcode from instruction bit 0 - 6 for a total len of 7 */
+    let opcode = {
+        let pos = env.alloc_scratch();
+        unsafe { env.bitmask(&instruction, 7, 0, pos) }
+    };
+    /* verify opcode is 7 bits */
+    env.range_check8(&opcode, 7);
```
Comment (dannywillems) [inline o1vm/src/interpreters/riscv32i/interpreter.rs:1102]: I think we should have a lookup table with the opcode. Not any 7 bits value is a correct opcode.

File: o1vm/src/interpreters/riscv32i/interpreter.rs [hunk 5]
```diff
@@ -1077,8 +1076,157 @@ pub fn interpret_rtype<Env: InterpreterEnv>(_env: &mut Env, _instr: RInstruction
...
+
+    /* decode and parse bits from the full 32 bit instruction in accordance with the Rtype riscV spec
```
Comment (dannywillems) [inline o1vm/src/interpreters/riscv32i/interpreter.rs:1104]:
```suggestion
    /* decode and parse bits from the full 32 bits instruction in accordance with the Rtype riscV spec
```

Comment (dannywillems) [inline o1vm/src/interpreters/riscv32i/interpreter.rs:1104]:
```suggestion
    /* decode and parse bits from the full 32 bit instruction in accordance with the Rtype RISC-V32i spec
```

**Review by dannywillems — APPROVED** [review]
Would be nice to have the changes in a follow-up. The lookup check on the opcode is important.

**Comment by dannywillems** [comment]
Conflicts should be solved.

File: o1vm/src/interpreters/riscv32i/mod.rs (hunk 2 above)
Comment (svv232) [reply o1vm/src/interpreters/riscv32i/mod.rs:4]: i just chose this value because it worked during testing, a size of 20 was too small.

**Pull request merged**